File: manejadorConsulta.py
from flask import Flask, jsonify, request, session, abort
import requests
import os
import traceback
from dotenv import load_dotenv

# Mensaje de error para consultas fallidas
ERROR_CONSULTA = "ERROR AL REALIZAR LA CONSULTA"

# Inicializa la aplicación Flask
app = Flask(__name__)
app.secret_key = os.urandom(24)  # Clave secreta para la sesión
session = {}  # Inicializa la sesión

# Carga las variables de entorno
load_dotenv()
api_key = os.getenv("AZURE_OPENAI_API_KEY")
api_endpoint = os.getenv("AZURE_OPENAI_API_ENDPOINT")
assistantModel = os.getenv("AZURE_OPENAI_API_MODEL")
version_api = os.getenv("AZURE_OPENAI_API_VERSION")

class ManejadorConsultas:
    
    def __init__(self) -> None:
        pass

    def mostrarConsulta(self, pregunta):
        try:
            # Verificación de variables
            if not api_key or not api_endpoint:
                raise ValueError("AZURE_OPENAI_API_KEY o AZURE_OPENAI_API_ENDPOINT no están configurados correctamente.")
            
            # Construcción de la URL
            url = f"{api_endpoint}/deployments/{assistantModel}/chat/completions?api-version={version_api}"
            print(f"URL construida: {url}")  # Para depuración

            # Headers
            headers = {
                "api-key": api_key,
                "Content-Type": "application/json"
            }

            # Obtener historial de la sesión
            if 'historial' not in session:
                session['historial'] = []

            # Agregar la nueva pregunta al historial
            session['historial'].append({"role": "user", "content": pregunta})

            # Cuerpo de la petición
            data = {
                "messages": session['historial'],
                "temperature": 0.5,
                "max_tokens": 1000,
                "top_p": 1,
                "frequency_penalty": 0,
                "presence_penalty": 0
            }

            # Realizar la petición
            response = requests.post(url, headers=headers, json=data)
            print(f"Response status code: {response.status_code}")  # Para depuración
            print(f"Response content: {response.text}")  # Para depuración

            # Verificar la respuesta
            if response.status_code == 200:
                response_data = response.json()
                respuesta = response_data['choices'][0]['message']['content'].strip()
                
                # Agregar la respuesta al historial
                session['historial'].append({"role": "assistant", "content": respuesta})
                
                return jsonify({"respuesta": respuesta})
            else:
                return jsonify({"error": f'Error al generar mensaje: {response.text}'}), response.status_code
        
        except ValueError as e:
            return jsonify({"error": f'Configuración incorrecta: {e}'}), 500
        
        except requests.RequestException as e:
            return jsonify({"error": f'Error al realizar la solicitud: {e}'}), 500
        
        except Exception as e:
            return jsonify({"error": f'Ocurrió un error inesperado: {e}'}), 500


# Versión anterior sin contexto de respuesta
"""
    def mostrarConsulta(self, pregunta):
        try:
            # Verificación de variables
            if not api_key or not api_endpoint:
                raise ValueError("AZURE_OPENAI_API_KEY o AZURE_OPENAI_API_ENDPOINT no están configurados correctamente.")
            
            # Construcción de la URL
            url = f"{api_endpoint}/deployments/{assistantModel}/chat/completions?api-version={version_api}"
           
            print(f"URL construida: {url}")  # Para depuración

            # Headers
            headers = {
                "api-key": api_key,
                "Content-Type": "application/json"
            }

            # Cuerpo de la petición
            data = {
                "messages": [
                    {"role": "user", "content": pregunta}
                ],
                "temperature": 0.5,
                "max_tokens": 1000,
                "top_p": 1,
                "frequency_penalty": 0,
                "presence_penalty": 0
            }

            # Realizar la petición
            response = requests.post(url, headers=headers, json=data)
            print(f"Response status code: {response.status_code}")  # Para depuración
            print(f"Response content: {response.text}")  # Para depuración

            # Verificar la respuesta
            if response.status_code == 200:
                response_data = response.json()
                answer = 'AI: ' + response_data['choices'][0]['message']['content'].strip()
                return jsonify({"respuesta": answer})
            else:
                return jsonify({"error": f'Error al generar mensaje: {response.text}'}), response.status_code
        
        except ValueError as e:
            return jsonify({"error": f'Configuración incorrecta: {e}'}), 500
        
        except requests.RequestException as e:
            return jsonify({"error": f'Error al realizar la solicitud: {e}'}), 500
        
        except Exception as e:
            return jsonify({"error": f'Ocurrió un error inesperado: {e}'}), 500
"""

File: BackEnd_En_enASPNET_conC#/RetoOpenAi/RetoOpenAi/Clases/ManejadorConsulta.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using System.Web.Http;
using static System.Net.WebRequestMethods;

namespace RetoOpenAI.Clases
{
    public class ManejadorConsultas
    {
        // Se declaran variables para almacenar configuraciones esenciales para la interacción
        // con la API de OpenAI de una manera centralizada y fácilmente configurable.

        private readonly string apiKey = ConfigurationManager.AppSettings["AZURE_OPENAI_API_KEY"];
        private readonly string apiEndpoint = ConfigurationManager.AppSettings["AZURE_OPENAI_API_ENDPOINT"];
        private readonly string assistantModel = ConfigurationManager.AppSettings["AZURE_OPENAI_API_MODEL"];
        private readonly string versionApi = ConfigurationManager.AppSettings["AZURE_OPENAI_API_VERSION"];

        private List<Tuple<string, string>> contexto = new List<Tuple<string, string>>();

        // constructor
        public ManejadorConsultas() { }

        //Método para enviar pregunta y recibir una respuesta
        public async Task<HttpResponseMessage> MostrarConsulta(string pregunta)
        {
            try
            {
                if (string.IsNullOrEmpty(apiKey) || string.IsNullOrEmpty(apiEndpoint))
                {
                    throw new Exception("AZURE_OPENAI_API_KEY o AZURE_OPENAI_API_ENDPOINT no están configurados correctamente.");
                }

                string url = $"{apiEndpoint}/deployments/{assistantModel}/chat/completions?api-version={versionApi}";

                using (var client = new HttpClient())
                {
                    client.DefaultRequestHeaders.Add("api-key", apiKey);

                    //  objeto anónimo
                    //  la información necesaria
                    //  para enviar una solicitud a la API de OpenAI.
                    var data = new
                    {
                        messages = contexto.Select(t => new { role = "user", content = t.Item1 }).Append(new { role = "user", content = pregunta }),
                        temperature = 0.5,
                        max_tokens = 1000,
                        top_p = 1,
                        frequency_penalty = 0,
                        presence_penalty = 0
                    };
                    // se convierte el objeto data a JSON
                    // Envía una solicitud HTTP POST a la URL especificada con el JSON en el cuerpo de la solicitud.
                    // Espera asincrónicamente la respuesta del servidor.
                    // Almacena la respuesta en la variable response.

                    var response = await client.PostAsJsonAsync(url, data);

                    if (response.IsSuccessStatusCode)   //  Verifica si el código de estado HTTP
                                                        //  de la respuesta indica éxito (códigos de estado en el rango 200-299).
                                                        // Procesamiento de la Respuesta en Caso de
                    {
                        // Se lee el contenido de la respuesta HTTP
                        // y lo deserializa en un objeto dinámico(dynamic),
                        // lo que permite acceder a las propiedades del JSON recibido
                        // sin necesidad de definir una estructura de datos específica.

                       var responseData = await response.Content.ReadAsAsync<dynamic>();


                        JObject jsonObject = new JObject();
                        string respuesta = responseData.choices[0].message.content;
                        jsonObject["respuesta"] = respuesta;
                        
                        // Se guarda el historial
                        contexto.Add(new Tuple<string, string>(pregunta, respuesta));

                        return new HttpResponseMessage(HttpStatusCode.OK)
                        {
                            //Content = new StringContent("{\"respuesta\": \"" + respuesta + "\"}")
                            Content = new StringContent(jsonObject.ToString())
                        };
                    }
                    else
                    {
                        return new HttpResponseMessage(HttpStatusCode.InternalServerError)
                        {
                            Content = new StringContent("{\"error\": \"" + $"Error al generar mensaje: {response.ReasonPhrase}" + "\"}")
                        };
                    }
                }
            }
            catch (Exception ex)
            {
                return new HttpResponseMessage(HttpStatusCode.InternalServerError)
                {
                    Content = new StringContent("{\"error\": \"" + $"Ocurrió un error: {ex.Message}" + "\"}")
                };
            }
        }
    }
}